presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java

```diff
@@ -246,19 +266,57 @@ private Estimate calculateNullsFraction(Map<String, PartitionStatistics> statist
                     if (partitionsWithStatisticsCount == 0) {
                         return OptionalDouble.empty();
                     }
-                    else {
-                        return OptionalDouble.of(totalPartitionsCount / partitionsWithStatisticsCount * nullsCount);
-                    }
+                    return OptionalDouble.of(totalPartitionsCount / partitionsWithStatisticsCount * nullsCount);
                 });
 
-        if (totalNullsCount.isValueUnknown() || totalRowsCount.isValueUnknown()) {
+        if (totalNullsCount.isValueUnknown()) {
             return Estimate.unknownValue();
         }
-        if (totalRowsCount.getValue() == 0.0) {
+        return new Estimate(totalNullsCount.getValue() / rowCount.getValue());
+    }
+
+    private Estimate calculateDataSize(Map<String, PartitionStatistics> statisticsSample, String columnName, Estimate rowCount)
+    {
+        if (rowCount.isValueUnknown()) {
+            return Estimate.unknownValue();
+        }
+
+        int knownPartitionCount = 0;
+        double knownRowCount = 0;
+        double knownDataSize = 0;
+
+        for (PartitionStatistics statistics : statisticsSample.values()) {
+            if (!statistics.getBasicStatistics().getRowCount().isPresent()) {
+                continue;
+            }
+            double partitionRowCount = statistics.getBasicStatistics().getRowCount().getAsLong();
+
+            HiveColumnStatistics partitionColumnStatistics = statistics.getColumnStatistics().get(columnName);
+            if (partitionColumnStatistics == null || !partitionColumnStatistics.getAverageColumnLength().isPresent()) {
+                continue;
+            }
+
+            double partitionNonNullCount = partitionRowCount - partitionColumnStatistics.getNullsCount().orElse(0);
+            if (partitionNonNullCount < 0) {
+                log.debug("null count bigger than row count in partition stats");
+                continue;
+            }
+            knownPartitionCount++;
+            knownRowCount += partitionRowCount;
+            // Note: average column length from Hive might not translate directly into internal data size
+            knownDataSize += partitionColumnStatistics.getAverageColumnLength().getAsDouble() * partitionNonNullCount;
+        }
+
+        if (knownPartitionCount == 0) {
+            return Estimate.unknownValue();
+        }
+
+        if (knownDataSize == 0) {
             return Estimate.zeroValue();
         }
 
-        return new Estimate(totalNullsCount.getValue() / totalRowsCount.getValue());
+        verify(knownRowCount > 0);
+        return new Estimate(knownDataSize / knownRowCount * rowCount.getValue());
```

Comment:
This should be `new Estimate(knownDataSize / nonNullKnownRowCount * totalKonNullRowCount);`

as `TableScanStatsRule#toSymbolStatistics` uses:
```
.setAverageRowSize(columnStatistics.getOnlyRangeColumnStatistics().getDataSize().getValue() / nonNullRowsCount)
```
as `RangeColumnStatistics#getDataSize` only relates to non-null rows.

Also see my comment: https://github.com/prestodb/presto/pull/11107#discussion_r208210852
I think this method can be much simplified to: https://github.com/starburstdata/presto-private/blob/0a47179b9d0df73724863fba226372a31c9355f5/presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java#L381

Reply:
https://github.com/prestodb/presto/pull/11107#discussion_r208280084

Reply:
Still:
```
This should be new Estimate(knownDataSize / nonNullKnownRowCount * totalKonNullRowCount);

as TableScanStatsRule#toSymbolStatistics uses:

.setAverageRowSize(columnStatistics.getOnlyRangeColumnStatistics().getDataSize().getValue() / nonNullRowsCount)

as RangeColumnStatistics#getDataSize only relates to non-null rows.
```
is a valid comment I think

Reply:
What we are doing here is a some kind of a linear "extrapolation"

Given the number of nulls distribution is the same (or similar) across the partition 

The

```
<know_data_size> * ( <total_non_null_row_count> / <known_non_null_row_count>)
```

is effectively the same as

```
<know_data_size> * ( <total_row_count> / <known_row_count> )
```

Since

```
<know_data_size> * ((<total_row_count> - (<total_row_count> * <null_fraction>)) / (<known_row_count> - (<known_row_count> * <null_fraction>))
```
=
```
<know_data_size> * ((<total_row_count> * (1 - <null_fraction>)) / (<known_row_count> * (1 - <null_fraction>))
```
=
```
<know_data_size> * (<total_row_count>  / <known_row_count>)
```